pytorch_lightning/utilities/device_dtype_mixin.py

```diff
@@ -50,6 +52,7 @@ def device(self, new_device: Union[str, torch.device]):
         # Necessary to avoid infinite recursion
         raise RuntimeError('Cannot set the device explicitly. Please use module.to(new_device).')
 
+    @parameter_validation
```

Comment:
We are trying not to use decorator as they are hard to debug and easy to forget. 
And I think there is already a PR pending for introducing a `on_post_move_to_device` callback.
I think it was introduced by @rohitgr7.

Reply:
this one is for models, `on_before/after_transfert_to_device` is for data. It's `on_before/after_batch_transfer`.

Reply:
I feel like this kind of check shall in a sanity check there is no need to execute in each call `to`

Reply:
@tchaton @Borda should we have this check without the decorator? move it somewhere else?

Reply:
@lezwon do we really need to run this check with every `to(...)` or can it be just for a single epoch, ale it can be overkill to wrap and check if it is empty...
so if this is not essential for this PR lets move it to separate PR and let this fix land asap 🐰

Reply:
cc: @PyTorchLightning/core-contributors

Reply:
Any chance the weight tying validation can happen within the TPU accelerator? and happen only after model_to_device is called?

Reply:
Yes, we should move it there in another PR.